### ISSUE TITLE
Meterpreter Stage Fails to use NTLM authentication proxy

### ISSUE THREAD

**Issue by Loki1980** [rapid7/metasploit-framework#5576]
Hi all, 

I'm having trouble using windows/meterpreter/reverse_https in an environment with a NTLM authentication proxy.
The initial connection works fine, but then, when the stage takes over, it is unable to connect, it tries a direct connection ( which gets blocked by the firewall ) to the host ignoring the proxy.

This is what I see from the metasploit console:

[*] Y.Y.Y.Y:46329 (UUID: 5d5ac57ec9aadc36/x86=1/windows=1/2015-06-19T13:40:01Z) Staging Native payload ...
[*] Meterpreter session 1 opened (X.X.X.X:443 -> Y.Y.Y.Y:46329) at 2015-06-22 11:02:05 +0200
[-] Meterpreter session 1 is not valid and will be closed
[*] Y.Y.Y.Y - Meterpreter session 1 closed.

If I manually set the advanced options PayloadProxyHost and PayloadProxyPort in the multihandler, then the stage tries to use the NTLM proxy, but it fails because it does not send the "Proxy-Connection: Keep-Alive" header, so it does not get the "Proxy-Authenticate" header with the challenge from the proxy.

Below an example of request/response :

CONNECT X.X.X.X:443 HTTP/1.1
Host: X.X.X.X:443
Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB4IIogAAAAAAAAAAAAAAAAAAAAAGAbEdAAAADw==

HTTP/1.0 407 Proxy Authentication Required
Server: squid/2.7.STABLE9
Date: Mon, 22 Jun 2015 10:03:24 GMT
Content-Type: text/html
Content-Length: 1169
X-Squid-Error: ERR_CACHE_ACCESS_DENIED 0
X-Cache: MISS from debian-proxy
X-Cache-Lookup: NONE from debian-proxy:3128
Via: 1.0 debian-proxy:3128 (squid/2.7.STABLE9)
Connection: close

If I change the config of the proxy in order to not perform NTLM authentication then the meterpreter stage works correctly.
I'm using the metasploit version v4.11.3-2015061001 [core:4.11.3.pre.2015061001 api:1.0.0]

**Comment by bcook-r7** [comment]
Hi @Loki1980 - I believe this is fixed in a later version of framework. The fixed packages for Kali linux will   be published later this week.

**Comment by OJ** [comment]
This should work with the latest, for sure. I'm not sure how far behind the Kali install is though. I know a few others have had similar issues, and the meterpreter binaries are out of date.

**Comment by bcook-r7** [comment]
Closing now that the latest kali image is released.

**Comment by Loki1980** [comment]
Hi guys, 

unfortunately I am experiencing the same issues even after the upgrade. The meterpreter/reverse_https works perfectly with the proxy only if it does not perform authentication: If (on the victim machine) I setup a local proxy (like zed attack proxy) that performs the NTLM authentication with the corporate proxy, and I set it as the default proxy, then the same meterpreter executable works perfectly, so the problem must be with the NTLM protocol.
Has anyone been able to get a meterpreter reverse connection in a secured environment where all the outbound traffic have to pass from a NTLM authenticating proxy ( in particular against a Squid proxy)?  
Any suggestions? I have tried to bypass the problem using a stageless payload, but it seems that they are not proxy-aware.
Also, if I use the windows/x64/meterpreter/reverse_https payload, I always get this error when sending the stage:


[*] X.X.X.X:45975 (UUID: ea8bce1ac1c37295/x64=3/windows=1/2015-06-24T08:02:00Z) Staging Native payload ...
[-] Exception handling request: undefined method `stage_payload' for #<#<Class:0x0000000982ed90>:0x0000000a3428d0>
[*] X.X.X.X:50841 (UUID: b59b962366765a8a/x64=3/windows=1/2015-06-24T08:02:31Z) Staging Native payload ...
[-] Exception handling request: undefined method `stage_payload' for #<#<Class:0x0000000982ed90>:0x0000000a3428d0>

Below the metasploit version and the relative packages I am using on kali:
metasploit v4.11.3-2015062101 [core:4.11.3.pre.2015062101 api:1.0.0]


metasploit-common:
  Installed: 1.0.2-0kali0
metasploit-framework:
  Installed: 4.11.3-2015062101-1kali0
metasploit:
  Installed: 4.11.3-2015062101-1kali0

**Comment by OJ** [comment]
OK, this isn't good. It _still_ looks like MSF is out of sync with the binaries. The missing `stage_payload` bug was fixed ages ago.

**Comment by hmoore-r7** [comment]
cc @shuckins-r7

**Comment by hmoore-r7** [comment]
@OJ stageless should definitely be proxy-aware

**Comment by bcook-r7** [comment]
It looks like the correct version of metasploit-payloads is with the latest Kali package. I'll do some more testing locally to see if we can understand better what is happening again.

**Comment by bcook-r7** [comment]
Agreed with @Loki1980 - I verified that unauthenticated proxy support works fine, and the Kali package is up to date now. So, this is not a packaging issue, and sounds more like a regular bug.

**Comment by bcook-r7** [comment]
Note, I did _not_ encounter the "undefined method `stage_payload'" issue using the latest kali linux package using windows/x64/meterpreter/reverse_https with an unauthenticated proxy. I also have not tested NTLM itself yet, so I do not know the two are related.

**Comment by Loki1980** [comment]
Hi all, 

just a little update: to do some more tests I installed a debian 8 from scratch and put metasploit on it from git.
With this installation (Framework: 4.11.0-dev Console  : 4.11.0-dev.15168) I don't get anymore the  missing stage_payload error, but I do still have all the other problems I wrote above.

**Comment by hmoore-r7** [comment]
@Loki1980 Thanks! Please open a new issue for any specific problems you encounter using the Git version of metasploit-framework.

**Comment by hmoore-r7** [comment]
The latest metasploit-framework and meterpreter payloads should be using any preconfigured IE proxy settings, including authentication, and this includes the stageless variants. If this isn't working, please file a new ticket with details with your setup.
